sharing_portal/views.py

```diff
@@ -1,226 +1,180 @@
 from django.template import loader
 from django.utils.html import strip_tags
 from django.utils import timezone
 from projects.models import Project
-from allocations.models import Allocation
-from rest_framework.renderers import JSONRenderer
-from rest_framework import serializers
+from projects.util import get_project_members
 from util.project_allocation_mapper import ProjectAllocationMapper
 from util.keycloak_client import KeycloakClient
 
 from .forms import (
     ArtifactForm,
-    ArtifactVersionForm,
     AuthorFormset,
     ShareArtifactForm,
-    ZenodoPublishFormset,
     RequestDaypassForm,
     ReviewDaypassForm,
 )
-from .models import Artifact, ArtifactVersion, Author, ShareTarget, DaypassRequest
-from .tasks import publish_to_zenodo
+from .models import DaypassRequest, DaypassProject
+from . import trovi
 from projects.views import (
     add_project_invitation,
     get_invite_url,
     get_project_membership_managers,
     is_membership_manager,
     manage_membership_in_scope,
 )
+from .zenodo import ZenodoClient
+
+from urllib.parse import urlencode
 
 import logging
+from datetime import datetime, timedelta
 
 LOG = logging.getLogger(__name__)
 
 SHARING_KEY_PARAM = "s"
 
 
-class ArtifactSerializer(serializers.ModelSerializer):
-    class Meta:
-        model = Artifact
-        fields = "__all__"
+def ensure_trovi_token(view_func):
+    def _wrapped_view(request, *args, **kwargs):
+        if (
+            request.session.get("trovi_token_expiration")
+            and datetime.now().timestamp() > request.session["trovi_token_expiration"]
+        ):
+            del request.session["trovi_token_expiration"]
+            del request.session["trovi_token"]
+
+        if not request.session.get("trovi_token"):
+            if request.session.get("oidc_access_token"):
+                try:
+                    response = trovi.get_token(
+                        request.session.get("oidc_access_token"),
+                        is_admin=request.user.is_staff,
+                    )
+                    request.session["trovi_token"] = response["access_token"]
+                    request.session["trovi_token_expiration"] = min(
+                        (
+                            datetime.now() + timedelta(seconds=response["expires_in"])
+                        ).timestamp(),
+                        request.session["oidc_token_expiration"],
+                    )
+                except trovi.TroviException:
+                    LOG.error("Error getting trovi token")
+            else:
+                # Set an empty token
+                request.session["trovi_token"] = ""
+        return view_func(request, *args, **kwargs)
+
+    return _wrapped_view
 
 
 def can_edit(request, artifact):
-    if artifact.created_by == request.user:
-        return True
-    if request.user.is_staff:
-        return True
-    return False
+    return _owns_artifact(request.user, artifact) or request.user.is_staff
```

Comment:
Does the staff flag mean anything anymore? Does having staff status translate to portal being allowed to generate an admin token for them?

Reply:
This assumes someone with `is_staff` is also a trovi admin. I can change `with_trovi_token` so to use the service admin token if user is staff.